### DISCUSSION TITLE
Memory management thread.

### DISCUSSION THREAD

**Discussion by MinekPo1** [Its-Jakey/SCPP-old-#6] [Ideas]
This is a thread to discuss memory management, both VM level and language level.

**Comment by MinekPo1** [comment]
On VM level:

Pointers:

- `getPointer` [var]: places the address of the variable into the `a` reg.

- `setPointer` [var]: sets the address of a variable to the value of the `a` reg.

- `setAt` [pointer]: sets the value at the position given by the pointer to the `a` reg.

- `setAtWithOffset` [pointer] [offset]: sets the value at the position given by the pointer + offset to the `a` reg.

Allocation:

- `free` [var]: remove the variable from the variable lookup table and mark the memory as free (maybe also clear it?)

- `malloc` [size]: (update) allocates [size] cells and sets the `a` reg to the address of the first cell

**Reply by Its-Jakey** [reply]
I'll start implementing these!

**Comment by MinekPo1** [comment]
Language level:

Note: all examples use the current typing system. See #5 for proposed changes to the type system.

Pointers and references:

```scpp
var a = 1;
var b = &a;
*b = 2;
println(a); //=> 2
var &c = &a;
c = 3;
println(a); //=> 3
```

`free` builtin:

```scpp
func free(*addr){
    var temp
    _ast_(
        "loadAtVar", addr,
        "setPointer", temp,
        "free", temp
    );
}
```

**Comment by Its-Jakey** [comment]
I replaced arrays with a much less buggy pointer-like system
